File: arrays_hashing/encode_and_decode_strings/encode_decode_strings_test.go
package encodedecode

import (
	"reflect"
	"testing"
)

func TestEncodeDecode(t *testing.T) {
	sol := &Solution{}

	tests := []struct {
		name  string
		input []string
	}{
		{
			name:  "Standard case",
			input: []string{"hello", "world"},
		},
		{
			name:  "Contains delimiter",
			input: []string{"hello#world", "test#123"},
		},
		{
			name:  "Empty strings",
			input: []string{"", "", ""},
		},
		{
			name:  "Empty slice",
			input: []string{},
		},
		{
			name:  "Long strings and spaces",
			input: []string{"The quick brown fox", "   ", "1234567890"},
		},
		{
			name:  "Unicode characters",
			input: []string{"你好", "🌍", "coding"},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			encoded := sol.Encode(tt.input)
			decoded := sol.Decode(encoded)

			if !reflect.DeepEqual(tt.input, decoded) {
				t.Errorf("Decode() = %v, want %v", decoded, tt.input)
			}
		})
	}
}

File: arrays_hashing/encode_and_decode_strings/encode_decode_strings_test.py
import pytest
from encode_decode_strings import encode, decode


@pytest.mark.parametrize(
    "input_list",
    [
        (["hello", "world"]),
        ([""]),
        (["", ""]),
        ([]),
        (["#", "##", "4#", "#abc"]),
        (["The quick brown fox", "jumps over", "the lazy dog"]),
        (["python", "🐍", "世界"]),
    ],
    ids=[
        "Standard",
        "SingleEmpty",
        "MultiEmpty",
        "EmptyList",
        "Delimiters",
        "Spaces",
        "Unicode",
    ],
)
def test_encode_decode(input_list):
    # 1. Call the standalone encode function
    encoded_str = encode(input_list)

    # 2. Call the standalone decode function
    decoded_list = decode(encoded_str)

    # 3. Assert they match
    assert decoded_list == input_list, (
        f"Failed! \n"
        f"Original: {input_list}\n"
        f"Encoded:  {encoded_str}\n"
        f"Decoded:  {decoded_list}"
    )
